presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/SetOperationMerge.java

```diff
@@ -55,58 +59,117 @@ public SetOperationMerge(SetOperationNode node, Context context, SetOperationNod
+     * Merge all matching source nodes. This method is assumed to be used only for associative set operations: UNION and INTERSECT.
      *
      * @return Merged plan node if applied.
      */
     public Optional<SetOperationNode> merge()
     {
+        checkState(node instanceof UnionNode || node instanceof IntersectNode, "unexpected node type: %s", node.getClass().getSimpleName());
+
         Lookup lookup = context.getLookup();
         List<PlanNode> sources = node.getSources().stream()
                 .flatMap(lookup::resolveGroup)
                 .collect(Collectors.toList());
 
-        // There must be one same source node at least.
-        if (sources.stream().noneMatch(node.getClass()::isInstance)) {
-            return Optional.empty();
-        }
-
         ImmutableListMultimap.Builder<Symbol, Symbol> newMappingsBuilder = ImmutableListMultimap.builder();
+        boolean resultIsDistinct = false;
+        boolean rewritten = false;
 
         for (int i = 0; i < sources.size(); i++) {
             PlanNode source = sources.get(i);
-            if (node.getClass().equals(source.getClass())) {
-                SetOperationNode setOperationNode = (SetOperationNode) source;
-                addMergedMappings(setOperationNode, i, newMappingsBuilder);
+
+            // Determine if set operations can be merged and whether the resulting set operation is quantified DISTINCT or ALL
+            Optional<Boolean> mergedQuantifier = mergedQuantifierIsDistinct(node, source);
+            if (mergedQuantifier.isPresent()) {
+                addMergedMappings((SetOperationNode) source, i, newMappingsBuilder);
+                resultIsDistinct |= mergedQuantifier.get();
+                rewritten = true;
             }
             else {
                 // Keep mapping as it is
                 addOriginalMappings(source, i, newMappingsBuilder);
             }
         }
 
-        return Optional.of(instantiator.create(node.getId(), newSources, newMappingsBuilder.build(), node.getOutputSymbols()));
+        if (!rewritten) {
+            return Optional.empty();
+        }
+        if (node instanceof UnionNode) {
+            return Optional.of(new UnionNode(node.getId(), newSources, newMappingsBuilder.build(), node.getOutputSymbols()));
+        }
+        return Optional.of(new IntersectNode(node.getId(), newSources, newMappingsBuilder.build(), node.getOutputSymbols(), resultIsDistinct));
+    }
+
+    /**
+     * Check if node and child are mergable based on their set operation type and quantifier.
+     * <p>
+     * For parent and child of type UNION, merge is always possible and the assumed quantifier is ALL, because UnionNode always represents UNION ALL.
+     * <p>
+     * For parent and child of type INTERSECT, merge is always possible:
+     * - if parent and child are both INTERSECT ALL, the resulting set operation is INTERSECT ALL
+     * - otherwise, the resulting set operation is INTERSECT DISTINCT
```

Comment:
This needs additional clarification. If the child is DISTINCT, that branch is guaranteed to have unique values, so at most one element of the other branch will be retained -- this is equivalent to just doing DISTINCT on the parent.

Reply:
I added this explanation.